third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -4333,38 +4513,48 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
             },
             1 => {
                 let (cand, subs, instantiation) = matching.remove(0);
-                let (_, _, result_type) = cand.get_signature();
+                let (_type_params, params, result_type) = cand.get_signature();
                 let result_type = result_type.instantiate(&instantiation);
+                let param_types = params
+                    .iter()
+                    .map(|param| param.get_type().instantiate(&instantiation))
+                    .collect_vec();
+                let ftype = Type::Fun(
+                    Box::new(Type::tuple(param_types.clone())),
+                    Box::new(result_type.clone()),
+                    AbilitySet::DEFINED_FUNCTIONS,
+                );
+
                 // Commit the candidate substitution to this expression translator.
                 self.subs = subs;
 
-                // Check result type against expected type.
-                let ty = self.check_type(loc, &result_type, expected_type, context);
-                // When the expected type of this call is an immutable reference while the actual return type is mutable
-                // the type of the call should be mutable,
-                // otherwise, the type info in the bytecode will be incorrect.
-                // the freeze operation below will make sure the expression after freeze is immutable
-                let id = if result_type.is_mutable_reference() && ty.is_immutable_reference() {
-                    if let Type::Reference(_, inner_ty) = ty {
-                        self.new_node_id_with_type_loc(
-                            &Type::Reference(ReferenceKind::Mutable, inner_ty),
+                if mask != 0u128 && !matches!(cand, AnyFunEntry::UserFun(_)) {
```

Comment:
A comment what you are doing here would be helpful. What magic number is 0u128?

I'm worried about this change and the complexity of that we support `_`. The MVP doesn't need `_`, it would have been SO much easier to leave the type checking mostly the same, and just derive the closure from the lambda without new function during lifting, after type checking. We would be already working at VM part.